zipline/pipeline/factors/factor.py

```diff
@@ -382,6 +391,73 @@ def mathfunc(self):
 )
 
 
+class summary_funcs(object):
+    """Namespace of functions meant to be used with DailySummary.
+    """
+
+    @staticmethod
+    def mean(a, missing_value):
+        return nanmean(a, axis=1)
+
+    @staticmethod
+    def stddev(a, missing_value):
+        return nanstd(a, axis=1)
+
+    @staticmethod
+    def max(a, missing_value):
+        return nanmax(a, axis=1)
+
+    @staticmethod
+    def min(a, missing_value):
+        return nanmin(a, axis=1)
+
+    @staticmethod
+    def median(a, missing_value):
+        return nanmedian(a, axis=1)
+
+    @staticmethod
+    def sum(a, missing_value):
+        return nansum(a, axis=1)
+
+    @staticmethod
+    def notnull_count(a, missing_value):
+        return (~is_missing(a, missing_value)).sum(axis=1)
+
+    names = {k for k in locals() if not k.startswith('_')}
+
+
+def summary_method(name):
```

Comment:
I'm a bit confused what's going on in this function, but that might be because of my lack of understanding of how `clsdict` eventually gets used.  So this method returns a function that returns a `DailySummary` factor?  I guess my question is why do we have to return the function rather than just the class directly.

Reply:
This function defines a method that gets added to the `Factor` class, and that method constructs a `DailySummary` instance.

My original draft of this code had a bunch of new `Factor` methods written conventionally, which looked like:

```python

    @expect_types(mask=(Filter, NotSpecifiedType))
    @float64_only
    def mean(self, mask=NotSpecified):
        """Create 1D factor computing the average value of ``self``.

        Returns
        -------
        mean : zipline.pipeline.Factor
        """
        return DailySummary(
            summary_funcs.mean,
            self,
            self.dtype,
            self.missing_value,
            mask=mask,
        )

    @expect_types(mask=(Filter, NotSpecifiedType))
    @float64_only
    def median(self, mask=NotSpecified):
        """Make a factor that computes the median value of ``self``.

        Returns
        -------
        median : zipline.pipeline.Factor
        """
        return DailySummary(
            summary_funcs.median,
            self,
            self.dtype,
            self.missing_value,
            mask=mask
        )

    @expect_types(mask=(Filter, NotSpecifiedType))
    @float64_only
    def sum(self, mask=NotSpecified):
        """Create 1D factor that sums the values of ``self``.

        Returns
        -------
        sum : zipline.pipeline.Factor
        """
        return DailySummary(
            summary_funcs.sum,
            self,
            self.dtype,
            self.missing_value,
            mask=mask,
        )
```
etc.

If you squint at all of those though, you can see that they all have the same structure, which is a method that returns `DailySummary({function}, self, self.dtype, self.missing_value, mask=mask)`. Rather than have that structure duplicated 7 times, I instead wrote a function constructs the method as a function of the name of the method it calls.

In the code that manipulates `clsdict` below, `clsdict` is the namespace of attributes of the `Factor` class, which includes all the function objects that act as methods of Factor. When we do

```python
    clsdict.update(
        {name: summary_method(name) for name in summary_funcs.names},
    )
```

that's equivalent to doing

```
mean = summary_method('mean')
median = summary_method('median')
max = summary_method('max')
```

which is in turn equivalent to doing:

```
def max(self, mask):
    ... body of function emitted by summary_method
```